File: src/skrecsys/recommendation/_slim.py
"""SLIM: one elastic-net regression per item, ported from Ferrari Dacrema's framework."""

import math
import numbers
import warnings
from typing import Any

import numpy as np
import scipy.sparse as sp
from numpy.typing import NDArray
from sklearn.exceptions import ConvergenceWarning

from skrecsys import _core
from skrecsys._typing import override
from skrecsys.indexing import SparseSpace
from skrecsys.recommendation._base import SimilarityRecommender, kernel_csr, kernel_indices
from skrecsys.recommendation._incremental import (
    IncrementalRecommenderMixin,
    affected_item_rows,
    remap_dense_square,
    replace_rows,
)


class SLIMElasticNet(IncrementalRecommenderMixin, SimilarityRecommender):
    """Sparse linear item-item model whose columns are elastic-net regressions [1]_.

    Item ``j`` is predicted from every other item by a linear model fitted on the
    interaction matrix ``R``: column ``j`` of the weight matrix minimizes

    ``1 / (2 n_users_) * ||R[:, j] - R[:, -j] w||^2 + alpha * l1_ratio * ||w||_1
    + 0.5 * alpha * (1 - l1_ratio) * ||w||^2``

    over ``w``, by default constrained to be non-negative [2]_. The L1 term is what
    makes the model sparse, and self-similarity is excluded rather than penalized, so
    the diagonal is zero by construction. The score of item ``j`` for user ``u`` is
    ``sum_i r_ui * w(i, j)`` over the items ``i`` of the user.

    This is a port of ``SLIMElasticNetRecommender`` from the evaluation framework of
    Ferrari Dacrema et al. [3]_, which fits each column with scikit-learn's
    :class:`~sklearn.linear_model.ElasticNet`; the parameters keep scikit-learn's names
    and meaning. The reference passes the whole interaction matrix as the design matrix
    once per item, while the kernel here forms the item Gram matrix ``R^T R`` once and
    runs coordinate descent against it, which is the same problem and the same solution.
    Two differences are visible in the fitted weights: the reference computes in float32
    and visits coordinates in a random order, so its weights agree with ours only to
    float32 precision; and it keeps ``min(nnz - 1, n_neighbors)`` weights per column,
    dropping the smallest one when a column has no more than ``n_neighbors`` of them,
    where this implementation keeps all ``n_neighbors``.

    Like :class:`~skrecsys.recommendation.EASE`, fitting needs ``O(n_items^2)`` memory
    for the Gram matrix, which is what bounds the usable catalog size.

    Parameters
    ----------
    alpha : float, default=1.0
        Strength of the elastic-net penalty. Must be positive; larger values give
        sparser, more shrunken columns.
    l1_ratio : float, default=0.1
        Mix of the two penalties, from 0 for ridge to 1 for lasso.
    n_neighbors : int, default=100
        Number of largest weights kept per column, the reference's ``topK``.
    positive : bool, default=True
        Constrain the weights to be non-negative, as the reference does by default.
    max_iter : int, default=100
        Maximum number of coordinate-descent sweeps per column.
    tol : float, default=1e-4
        Stop a column once its duality gap falls below ``tol * ||R[:, j]||^2``.
    n_jobs : int or None, default=None
        Number of threads used to fit. ``None`` or ``-1`` uses all cores.
    index : None, str or VectorIndex, default=None
        Approximate index used by ``recommend``. ``None`` scores every candidate
        exactly; ``"hnsw"`` or a configured :class:`~skrecsys.indexing.HNSW` walks a
        graph instead. See :mod:`skrecsys.indexing`, and note that the exact path here
        is already an inverted-index scan that never touches an item nobody reached --
        so whether a graph beats it is a question for ``benchmarks/indexes.py``.

    Attributes
    ----------
    gram_ : ndarray of shape (n_items_, n_items_)
        The item Gram matrix every column's regression is solved from. Present only once
        ``partial_fit`` has been used.
    similarity_ : scipy.sparse.csr_array of shape (n_items_, n_items_)
        Item-item weights with a zero diagonal; entry ``(i, j)`` is the contribution of
        item ``i`` to the score of item ``j``, so column ``j`` holds one fitted
        regression.
    n_unconverged_ : int
        Number of columns that reached ``max_iter`` without meeting ``tol``. A warning
        is raised during ``fit`` when this is nonzero.

    Notes
    -----
    ``partial_fit`` keeps ``gram_``, updates it by the rank the batch has, and re-solves
    only the columns the batch can reach, warm-starting each from the weights it already
    had. It is an approximation, not a reproduction of ``fit`` on every batch
    concatenated: coordinate descent stops at a tolerance rather than at the optimum, so
    a warm start lands somewhere else inside it, and the l1 penalty scales with the
    number of users, so every new user moves the optimum of every column -- the
    unreachable ones included. Keeping ``gram_`` also doubles the model's memory, which
    is already quadratic in the catalog.

    References
    ----------
    .. [1] X. Ning and G. Karypis, "SLIM: Sparse Linear Methods for Top-N Recommender
       Systems", ICDM 2011. https://doi.org/10.1109/ICDM.2011.134
    .. [2] M. Levy and K. Jack, "Efficient Top-N Recommendation by Linear Regression",
       LSRS 2013, which replaces SLIM's bound-constrained solver with elastic net.
    .. [3] M. Ferrari Dacrema et al., "Are We Really Making Much Progress? A Worrying
       Analysis of Recent Neural Recommendation Approaches", RecSys 2019.
       https://github.com/MaurizioFD/RecSys2019_DeepLearning_Evaluation

    Examples
    --------
    >>> from skrecsys.recommendation import SLIMElasticNet
    >>> X = [["u1", "a"], ["u1", "b"], ["u2", "a"], ["u2", "b"], ["u3", "a"]]
    >>> rec = SLIMElasticNet(alpha=0.01).fit(X)
    >>> rec.recommend(["u3"], n_recommendations=1)[0].tolist()
    [['b']]
    """

    def __init__(
        self,
        alpha: float = 1.0,
        l1_ratio: float = 0.1,
        n_neighbors: int = 100,
        *,
        positive: bool = True,
        max_iter: int = 100,
        tol: float = 1e-4,
        n_jobs: int | None = None,
        index: Any = None,
    ) -> None:
        self.alpha = alpha
        self.l1_ratio = l1_ratio
        self.n_neighbors = n_neighbors
        self.positive = positive
        self.max_iter = max_iter
        self.tol = tol
        self.n_jobs = n_jobs
        self.index = index

    _incremental_state_ = (("similarity_", "item_item"),)

    @override
    def _fit(self, interactions: sp.csr_array) -> None:
        n_threads = self._check_params()
        n_items = interactions.shape[1]
        gram = _core.item_gram(*kernel_csr(interactions), n_items, n_threads)
        columns = self._solve(gram, interactions.shape[0], None, None, n_threads, n_items)
        # The kernel solves one column per row, so row j holds the model of item j.
        self.similarity_ = sp.csr_array(columns.T)
        # A plain `fit` keeps the memory it always did. The Gram matrix is dense and
        # quadratic in the catalog, and only an incremental fit has any use for it.
        self.__dict__.pop("gram_", None)

    @override
    def _remap(
        self,
        *,
        user_perm: NDArray[np.intp],
        item_perm: NDArray[np.intp],
        n_users: int,
        n_items: int,
    ) -> None:
        super()._remap(user_perm=user_perm, item_perm=item_perm, n_users=n_users, n_items=n_items)
        if "gram_" in self.__dict__:
            # An item the model has not seen has an all-zero Gram row and column, which
            # is what `remap_dense_square` leaves behind.
            self.gram_ = remap_dense_square(self.gram_, item_perm, n_items)

    @override
    def _partial_fit(
        self,
        interactions: sp.csr_array,
        *,
        delta: sp.csr_array,
        new_user_indices: NDArray[np.intp],
        new_item_indices: NDArray[np.intp],
        touched_user_indices: NDArray[np.intp],
        touched_item_indices: NDArray[np.intp],
    ) -> None:
        """Re-solve the columns the batch can reach, warm-started, over an updated Gram.

        The Gram matrix is a sum of the users' outer products, so the batch changes it by
        removing what the touched users used to contribute and adding what they
        contribute now -- a rank-limited update rather than another pass over the
        interactions.

        This is the one estimator here whose ``partial_fit`` is **not** what ``fit`` on
        every batch concatenated would give, and for two reasons worth stating. The
        descent stops at a duality gap rather than at the exact optimum, so a warm start
        lands at a different point inside the same tolerance. And the l1 penalty is
        scaled by the number of users, so every new user moves every column's optimum --
        including the columns the batch cannot reach, which are not re-solved.
        """
        del new_user_indices, new_item_indices, touched_item_indices
        n_threads = self._check_params()
        n_users, n_items = interactions.shape
        if "gram_" not in self.__dict__:
            # The model was `fit`, not `partial_fit`: there is no Gram matrix to update,
            # and one pass over the interactions builds the one every later call keeps.
            self.gram_ = _core.item_gram(*kernel_csr(interactions), n_items, n_threads)
            columns = self._solve(self.gram_, n_users, None, None, n_threads, n_items)
            self.similarity_ = sp.csr_array(columns.T)
            return

        after = sp.csr_array(interactions[touched_user_indices])
        before = after - sp.csr_array(delta[touched_user_indices])
        # Kept sparse on the way in: the difference of the two outer products has only
        # the touched users' items in it, and densifying it would cost the catalog
        # squared twice over.
        difference = sp.coo_array(sp.csr_array(after.T @ after) - sp.csr_array(before.T @ before))
        self.gram_[difference.row, difference.col] += difference.data

        moved = np.unique(after.indices)
        targets = affected_item_rows(interactions, moved.astype(np.intp))
        stored = sp.csr_array(self.similarity_.T)
        block = self._solve(
            self.gram_, n_users, targets, sp.csr_array(stored[targets]), n_threads, n_items
        )
        self.similarity_ = sp.csr_array(replace_rows(stored, targets, block).T)

    def _solve(
        self,
        gram: NDArray[np.float64],
        n_users: int,
        targets: NDArray[np.intp] | None,
        warm: sp.csr_array | None,
        n_threads: int,
        n_items: int,
    ) -> sp.csr_array:
        """Solve the elastic net of each target column; row ``t`` is target ``t``."""
        (indptr, indices, data), unconverged = _core.slim_elasticnet_weights_from_gram(
            gram,
            float(n_users),
            float(self.alpha),
            float(self.l1_ratio),
            bool(self.positive),
            int(self.max_iter),
            float(self.tol),
            int(self.n_neighbors),
            n_threads,
            None if targets is None else kernel_indices(targets),
            *(
                (None, None, None)
                if warm is None
                else (
                    kernel_indices(warm.indptr),
                    kernel_indices(warm.indices),
                    np.ascontiguousarray(warm.data, dtype=np.float64),
                )
            ),
        )
        solved = n_items if targets is None else len(targets)
        self.n_unconverged_ = int(unconverged)
        if unconverged:
            warnings.warn(
                f"{unconverged} of {solved} columns did not converge within "
                f"max_iter={self.max_iter}; raise max_iter, raise tol or raise alpha.",
                ConvergenceWarning,
                # _fit is called by fit, which the caller called.
                stacklevel=4,
            )
        return sp.csr_array((data, indices, indptr), shape=(solved, n_items), dtype=np.float64)

    @override
    def _index_space(self) -> SparseSpace:
        # Row `j` of `_neighbors_of_items` holds exactly the weights that make up item
        # `j`'s score, which is the item vector, and it is already built and cached.
        return SparseSpace(sp.csr_array(self._neighbors_of_items()))

    @override
    def _index_queries(self, user_indices: NDArray[np.intp]) -> sp.csr_array:
        return sp.csr_array(self.interactions_[user_indices])

    @override
    def _check_params(self) -> int:
        """Validate parameters and return the thread count for the kernel (0 = all)."""
        if (
            not isinstance(self.alpha, numbers.Real)
            or not math.isfinite(float(self.alpha))
            or self.alpha <= 0
        ):
            raise ValueError(f"alpha must be a finite real number > 0, got {self.alpha!r}.")
        if not isinstance(self.l1_ratio, numbers.Real) or not 0 <= float(self.l1_ratio) <= 1:
            raise ValueError(f"l1_ratio must be a real number in [0, 1], got {self.l1_ratio!r}.")
        if not isinstance(self.n_neighbors, numbers.Integral) or self.n_neighbors < 1:
            raise ValueError(f"n_neighbors must be an integer >= 1, got {self.n_neighbors!r}.")
        if not isinstance(self.positive, bool):
            raise ValueError(f"positive must be a bool, got {self.positive!r}.")
        if not isinstance(self.max_iter, numbers.Integral) or self.max_iter < 1:
            raise ValueError(f"max_iter must be an integer >= 1, got {self.max_iter!r}.")
        if (
            not isinstance(self.tol, numbers.Real)
            or not math.isfinite(float(self.tol))
            or self.tol <= 0
        ):
            raise ValueError(f"tol must be a finite real number > 0, got {self.tol!r}.")
        if self.n_jobs is None or self.n_jobs == -1:
            return 0
        if not isinstance(self.n_jobs, numbers.Integral) or self.n_jobs < 1:
            raise ValueError(f"n_jobs must be None, -1 or an integer >= 1, got {self.n_jobs!r}.")
        return int(self.n_jobs)

File: rust/kernels/src/slim.rs
//! SLIM with elastic-net weights: one sparse linear regression per item, fitted by
//! coordinate descent (X. Ning and G. Karypis, "SLIM: Sparse Linear Methods for Top-N
//! Recommender Systems", ICDM 2011; M. Levy and K. Jack, "Efficient Top-N
//! Recommendation by Linear Regression", LSRS 2013).
//!
//! Item `j` is regressed on every other item, `min_w 1/(2 m) ||r_j - R_-j w||^2 +
//! alpha l1_ratio ||w||_1 + alpha (1 - l1_ratio) / 2 ||w||^2` over `m` users, with `w`
//! usually constrained to be non-negative. Column `j` of the weight matrix is that
//! solution, and the diagonal is zero by construction.
//!
//! The reference implementation (M. Ferrari Dacrema et al.,
//! `SLIM_ElasticNet/SLIMElasticNetRecommender.py`) hands each column to scikit-learn's
//! `ElasticNet` with the whole interaction matrix as the design matrix, which re-reads
//! `R` once per item. Every solve only ever touches `R` through `R^T R`, so this kernel
//! forms the item Gram matrix once, shared with EASE, and runs the same coordinate
//! descent scikit-learn runs when `precompute=True`: for target `j` the subproblem is
//! `min_w 1/2 w^T G w - G_j^T w + l1 ||w||_1 + l2 / 2 ||w||^2` with row and column `j`
//! of `G` left out. Zeroing the target column, as the reference does, leaves `G[j][j]`
//! as its only trace, and that is `||r_j||^2`, which the stopping rule needs.
//!
//! Columns are solved in parallel and pruned to their `k` largest weights. Memory is
//! quadratic in the number of items, like EASE.

use crate::ease::gram;
use std::sync::atomic::{AtomicUsize, Ordering};

use crate::sparse::{Csr, CsrOwned};
use crate::vectors::axpy;

/// The per-item elastic net, in scikit-learn's parameterization.
pub struct ElasticNet {
    pub alpha: f64,
    pub l1_ratio: f64,
    pub positive: bool,
    pub max_iter: usize,
    pub tol: f64,
}

/// Per-item elastic-net weights for a users x items CSR matrix `r`, pruned to `k`.
///
/// Row `j` of the result holds the `k` largest weights of the regression that predicts
/// item `j`, that is column `j` of the SLIM weight matrix; entry `j` itself is never
/// stored. Rows are solved in parallel on the current rayon pool, sorted by item index,
/// and exact zeros are dropped. Returns the number of columns that hit `max_iter`
/// without converging alongside the weights.
pub fn similarity(r: &Csr, enet: &ElasticNet, k: usize) -> (CsrOwned, usize) {
    let g = gram(r);
    similarity_from_gram(&g, r.n_cols, r.n_rows as f64, None, None, enet, k)
}

/// [`similarity`] from an item Gram matrix that is already in hand.
///
/// `g` is the dense row-major `n x n` [`gram`] and `m` the number of users, which the
/// penalties scale by. `columns` restricts the solve to those targets, in that order,
/// and `warm` gives each of them a starting point: row `t` of `warm` holds the
/// coefficients to start target `columns[t]` from. Both `None` reproduces
/// [`similarity`] exactly.
///
/// An incremental fit uses all three: it keeps the Gram matrix and updates it by the
/// rank the batch has, re-solves only the columns the batch can reach, and starts each
/// of those from the solution it already had. Coordinate descent stops at a duality
/// gap rather than at the exact optimum, so a warm start lands at a different point
/// inside the same tolerance -- and the columns that are not re-solved keep an optimum
/// that has moved. This is an approximation of a full fit and does not claim otherwise.
pub fn similarity_from_gram(
    g: &[f64],
    n: usize,
    m: f64,
    columns: Option<&[usize]>,
    warm: Option<&Csr>,
    enet: &ElasticNet,
    k: usize,
) -> (CsrOwned, usize) {
    // scikit-learn folds the `1 / (2 m)` of the least-squares term into the penalties.
    let l1 = enet.alpha * enet.l1_ratio * m;
    let l2 = enet.alpha * (1.0 - enet.l1_ratio) * m;

    let unconverged = AtomicUsize::new(0);
    let init = || Solver::new(n);
    let solve = |solver: &mut Solver, at: usize, j: usize, out: &mut Vec<(usize, f64)>| {
        let start = warm.map_or(0..0, |w| w.indptr[at]..w.indptr[at + 1]);
        let warm_row = warm.map(|w| (&w.indices[start.clone()], &w.data[start]));
        if !solver.solve(g, n, j, l1, l2, enet, warm_row) {
            unconverged.fetch_add(1, Ordering::Relaxed);
        }
        top_k(&solver.w, j, k, out);
    };
    let weights = match columns {
        Some(columns) => CsrOwned::build_rows(columns, init, |solver, j, out| {
            let at = columns.binary_search(&j).expect("a row that was asked for");
            solve(solver, at, j, out);
        }),
        None => CsrOwned::build(n, init, |solver, j, out| solve(solver, j, j, out)),
    };
    (weights, unconverged.into_inner())
}

/// Reusable buffers for one column's coordinate descent.
struct Solver {
    /// Coefficients of the current column.
    w: Vec<f64>,
    /// `G w`, kept up to date so that a coordinate update costs one row of `G`.
    h: Vec<f64>,
}

impl Solver {
    fn new(n: usize) -> Self {
        Self {
            w: vec![0.0; n],
            h: vec![0.0; n],
        }
    }

    /// Fit column `target`, leaving the coefficients in `self.w`; `true` if it converged.
    ///
    /// Cyclic coordinate descent with the soft-thresholding update and the duality-gap
    /// stopping rule of scikit-learn's `enet_coordinate_descent_gram`. The reference
    /// leaves `selection='random'` on, which visits the same coordinates in a random
    /// order; both converge to the same optimum of a strictly convex problem.
    #[allow(clippy::too_many_arguments)]
    fn solve(
        &mut self,
        g: &[f64],
        n: usize,
        target: usize,
        l1: f64,
        l2: f64,
        enet: &ElasticNet,
        warm: Option<(&[i64], &[f64])>,
    ) -> bool {
        self.w.fill(0.0);
        self.h.fill(0.0);
        if let Some((indices, values)) = warm {
            // A coordinate the descent will not visit must not keep a warm value, or it
            // would survive into the output without ever having been solved for.
            for (&ii, &value) in indices.iter().zip(values) {
                let ii = ii as usize;
                if ii != target && g[ii * n + ii] != 0.0 {
                    self.w[ii] = value;
                }
            }
            for &ii in indices {
                let ii = ii as usize;
                if self.w[ii] != 0.0 {
                    axpy(&mut self.h, &g[ii * n..(ii + 1) * n], self.w[ii]);
                }
            }
        }
        // `G` is symmetric, so its column `target` is the contiguous row `target`.
        let q = &g[target * n..(target + 1) * n];
        let y_norm2 = q[target];
        if y_norm2 == 0.0 {
            // An item nobody interacted with: the target is the zero vector.
            return true;
        }

        for iteration in 0..enet.max_iter {
            let (mut w_max, mut d_w_max) = (0.0_f64, 0.0_f64);
            for ii in 0..n {
                let g_ii = g[ii * n + ii];
                if ii == target || g_ii == 0.0 {
                    continue;
                }
                let row = &g[ii * n..(ii + 1) * n];
                let w_ii = self.w[ii];
                if w_ii != 0.0 {
                    axpy(&mut self.h, row, -w_ii);
                }
                let rho = q[ii] - self.h[ii];
                self.w[ii] = if enet.positive && rho <= 0.0 {
                    0.0
                } else {
                    rho.signum() * (rho.abs() - l1).max(0.0) / (g_ii + l2)
                };
                if self.w[ii] != 0.0 {
                    axpy(&mut self.h, row, self.w[ii]);
                }
                d_w_max = d_w_max.max((self.w[ii] - w_ii).abs());
                w_max = w_max.max(self.w[ii].abs());
            }
            let last = iteration + 1 == enet.max_iter;
            if w_max == 0.0 || d_w_max / w_max < enet.tol || last {
                let gap = self.duality_gap(q, target, l1, l2, y_norm2, enet.positive);
                if gap < enet.tol * y_norm2 {
                    return true;
                }
                if last {
                    return false;
                }
            }
        }
        true
    }

    /// The elastic-net duality gap at the current `w`, as scikit-learn computes it.
    ///
    /// The dual point is the residual rescaled to be feasible, `const = l1 /
    /// max(|G w - q|)` when that exceeds one, and the gap is the difference between the
    /// primal objective and the resulting dual objective. Index `target` is skipped
    /// throughout: the reference zeroes that column of the design matrix, which zeroes
    /// its entry of `q` and of `G w` together.
    fn duality_gap(
        &self,
        q: &[f64],
        target: usize,
        l1: f64,
        l2: f64,
        y_norm2: f64,
        positive: bool,
    ) -> f64 {
        let (mut q_dot_w, mut h_dot_w, mut w_norm1, mut w_norm2) = (0.0, 0.0, 0.0, 0.0);
        let mut dual_norm = 0.0_f64;
        for (i, (&q_i, (&h_i, &w_i))) in q.iter().zip(self.h.iter().zip(&self.w)).enumerate() {
            if i == target {
                continue;
            }
            q_dot_w += q_i * w_i;
            h_dot_w += h_i * w_i;
            w_norm1 += w_i.abs();
            w_norm2 += w_i * w_i;
            let xta = q_i - h_i - l2 * w_i;
            dual_norm = dual_norm.max(if positive { xta } else { xta.abs() });
        }

        let r_norm2 = y_norm2 + h_dot_w - 2.0 * q_dot_w;
        let (scale, mut gap) = if dual_norm > l1 {
            let scale = l1 / dual_norm;
            (scale, 0.5 * (r_norm2 + r_norm2 * scale * scale))
        } else {
            (1.0, r_norm2)
        };
        gap += l1 * w_norm1 - scale * y_norm2 + scale * q_dot_w;
        gap + 0.5 * l2 * (1.0 + scale * scale) * w_norm2
    }
}

/// The `k` largest weights of a fitted column, sorted by item index.
///
/// Ranks by descending weight and breaks ties by ascending item index; the reference
/// ranks by descending weight too, with an unstable sort.
fn top_k(w: &[f64], target: usize, k: usize, out: &mut Vec<(usize, f64)>) {
    let before = out.len();
    out.extend(
        w.iter()
            .enumerate()
            .filter(|&(i, &value)| i != target && value != 0.0)
            .map(|(i, &value)| (i, value)),
    );
    if out.len() - before > k {
        crate::knn::keep_best(out, before, k);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Warm starting must not change where a tight solve lands: the problem is strictly
    /// convex, so both starting points converge on the same optimum.
    #[test]
    fn a_warm_start_reaches_the_same_optimum() {
        use crate::sparse::testing::{Owned, pseudo_random_dense, row};
        let dense = pseudo_random_dense(60, 14);
        let r = Owned::from_dense(&dense);
        let enet = ElasticNet {
            alpha: 0.05,
            l1_ratio: 0.5,
            positive: true,
            max_iter: 2000,
            tol: 1e-12,
        };
        let (cold, _) = similarity(&r.csr(), &enet, 14);
        let g = gram(&r.csr());
        let wanted = [2usize, 5, 9];
        let warm = Owned::from_rows(
            wanted.iter().map(|&j| row(&cold, j)).collect::<Vec<_>>(),
            14,
        );
        let (warmed, _) =
            similarity_from_gram(&g, 14, 60.0, Some(&wanted), Some(&warm.csr()), &enet, 14);
        for (at, &j) in wanted.iter().enumerate() {
            let (got, want) = (row(&warmed, at), row(&cold, j));
            assert_eq!(got.len(), want.len(), "column {j}");
            for ((gj, gv), (wj, wv)) in got.iter().zip(&want) {
                assert_eq!(gj, wj, "column {j}");
                assert!((gv - wv).abs() < 1e-8, "column {j}: {gv} vs {wv}");
            }
        }
    }

    /// Asking for every column with no warm start is the unrestricted kernel, which is
    /// the guarantee that lets `fit` and `partial_fit` share one solver.
    #[test]
    fn every_column_cold_is_the_unrestricted_fit() {
        use crate::sparse::testing::{Owned, pseudo_random_dense};
        let dense = pseudo_random_dense(40, 11);
        let r = Owned::from_dense(&dense);
        let enet = ElasticNet {
            alpha: 0.05,
            l1_ratio: 0.5,
            positive: true,
            max_iter: 100,
            tol: 1e-6,
        };
        let all: Vec<usize> = (0..11).collect();
        let g = gram(&r.csr());
        let (full, a) = similarity(&r.csr(), &enet, 11);
        let (part, b) = similarity_from_gram(&g, 11, 40.0, Some(&all), None, &enet, 11);
        assert_eq!(
            (full.indptr, full.indices, full.data, a),
            (part.indptr, part.indices, part.data, b)
        );
    }

    use crate::sparse::testing::{Owned, pseudo_random_dense, row};

    fn enet(alpha: f64, l1_ratio: f64, positive: bool) -> ElasticNet {
        ElasticNet {
            alpha,
            l1_ratio,
            positive,
            max_iter: 1000,
            tol: 1e-10,
        }
    }

    /// `(G w - q)_i`, the gradient of the least-squares part of the subproblem.
    fn gradient(g: &[f64], n: usize, target: usize, w: &[f64], i: usize) -> f64 {
        let dot: f64 = (0..n)
            .filter(|&c| c != target)
            .map(|c| g[i * n + c] * w[c])
            .sum();
        dot - g[target * n + i]
    }

    /// Assert the first-order optimality conditions of the elastic net at `w`.
    ///
    /// The subproblem is strictly convex, so they pin the solution without needing a
    /// second implementation to compare against: a nonzero coefficient sits where the
    /// gradient balances the penalties, and a zero one where the `l1` term dominates.
    fn assert_optimal(
        g: &[f64],
        n: usize,
        target: usize,
        w: &[f64],
        l1: f64,
        l2: f64,
        positive: bool,
    ) {
        for i in (0..n).filter(|&i| i != target && g[i * n + i] != 0.0) {
            let grad = gradient(g, n, target, w, i) + l2 * w[i];
            if w[i] != 0.0 {
                assert!(w[i] > 0.0 || !positive, "coefficient {i} is negative");
                let residual = grad + l1 * w[i].signum();
                assert!(residual.abs() < 1e-6, "coefficient {i}: {residual}");
            } else if positive {
                assert!(grad + l1 > -1e-6, "zero coefficient {i}: {grad}");
            } else {
                assert!(grad.abs() < l1 + 1e-6, "zero coefficient {i}: {grad}");
            }
        }
    }

    fn fit(dense: &[Vec<f64>], enet: &ElasticNet, k: usize) -> (CsrOwned, Vec<f64>, usize) {
        let owned = Owned::from_dense(dense);
        let (weights, unconverged) = similarity(&owned.csr(), enet, k);
        (weights, gram(&owned.csr()), unconverged)
    }

    #[test]
    fn every_column_solves_its_elastic_net() {
        let dense = pseudo_random_dense(60, 14);
        let (n, m) = (14, 60.0);
        let params = enet(0.05, 0.3, true);
        let (weights, g, unconverged) = fit(&dense, &params, n);
        assert_eq!(unconverged, 0);
        let (l1, l2) = (params.alpha * params.l1_ratio * m, params.alpha * 0.7 * m);

        for target in 0..n {
            let mut w = vec![0.0; n];
            for (i, value) in row(&weights, target) {
                assert_ne!(i, target, "the diagonal must stay zero");
                w[i] = value;
            }
            assert_optimal(&g, n, target, &w, l1, l2, true);
        }
    }

    #[test]
    fn signed_weights_solve_the_unconstrained_problem() {
        let dense = pseudo_random_dense(50, 12);
        let (n, m) = (12, 50.0);
        let params = enet(0.02, 0.5, false);
        let (weights, g, _) = fit(&dense, &params, n);
        let (l1, l2) = (params.alpha * params.l1_ratio * m, params.alpha * 0.5 * m);

        let mut negatives = 0;
        for target in 0..n {
            let mut w = vec![0.0; n];
            for (i, value) in row(&weights, target) {
                w[i] = value;
                negatives += usize::from(value < 0.0);
            }
            assert_optimal(&g, n, target, &w, l1, l2, false);
        }
        assert!(
            negatives > 0,
            "the unconstrained fit found no negative weight"
        );
    }

    #[test]
    fn a_larger_alpha_shrinks_the_weights() {
        let dense = pseudo_random_dense(80, 15);
        let sum = |alpha| {
            let (weights, _, _) = fit(&dense, &enet(alpha, 0.3, true), 15);
            (weights.data.iter().sum::<f64>(), weights.data.len())
        };
        let (weak, weak_nnz) = sum(0.005);
        let (strong, strong_nnz) = sum(0.5);
        assert!(strong < weak, "{strong} vs {weak}");
        assert!(strong_nnz < weak_nnz, "{strong_nnz} vs {weak_nnz}");
    }

    #[test]
    fn keeps_the_k_largest_weights_of_each_column() {
        let dense = pseudo_random_dense(70, 16);
        let params = enet(0.005, 0.2, true);
        let (full, _, _) = fit(&dense, &params, 16);
        let k = 3;
        let (pruned, _, _) = fit(&dense, &params, k);
        for target in 0..16 {
            let kept = row(&pruned, target);
            assert!(kept.len() <= k, "column {target} kept {}", kept.len());
            let smallest = kept.iter().map(|&(_, v)| v).fold(f64::MAX, f64::min);
            for (i, value) in row(&full, target) {
                if !kept.iter().any(|&(c, _)| c == i) {
                    assert!(value <= smallest, "column {target} dropped {value}");
                }
            }
        }
    }

    #[test]
    fn reports_columns_that_hit_the_iteration_limit() {
        let dense = pseudo_random_dense(60, 14);
        let params = ElasticNet {
            alpha: 1e-4,
            l1_ratio: 0.1,
            positive: true,
            max_iter: 1,
            tol: 1e-12,
        };
        let (_, _, unconverged) = fit(&dense, &params, 14);
        assert!(unconverged > 0);
    }

    #[test]
    fn an_item_without_interactions_has_no_weights() {
        // Item 2 is in the catalog but nobody interacted with it.
        let dense = vec![vec![1.0, 1.0, 0.0], vec![1.0, 2.0, 0.0]];
        let (weights, _, unconverged) = fit(&dense, &enet(0.001, 0.5, true), 3);
        assert_eq!(unconverged, 0);
        assert_eq!(row(&weights, 2), vec![]);
        // Nor can it be a neighbour of anything else.
        assert!(!weights.indices.contains(&2));
    }

    #[test]
    fn thread_count_does_not_change_the_result() {
        let dense = pseudo_random_dense(200, 40);
        let owned = Owned::from_dense(&dense);
        let run = |threads| {
            rayon::ThreadPoolBuilder::new()
                .num_threads(threads)
                .build()
                .expect("pool")
                .install(|| similarity(&owned.csr(), &enet(0.01, 0.3, true), 7).0)
        };
        let (a, b) = (run(1), run(4));
        assert_eq!(a.indptr, b.indptr);
        assert_eq!(a.indices, b.indices);
        assert_eq!(a.data, b.data);
    }
}
